src/measure_theory/ae_eq_fun.lean

```diff
@@ -0,0 +1,371 @@
+
+protected lemma add_left_neg : ∀ (a : α →ₘ γ), -a + a = 0 :=
+by rintros ⟨a⟩; exact quotient.sound (univ_mem_sets' $ assume a, add_left_neg _)
+
+instance : add_group (α →ₘ γ) :=
+{ neg          := ae_eq_fun.neg,
+  add_left_neg := ae_eq_fun.add_left_neg,
+  .. ae_eq_fun.add_monoid
+ }
+
+end add_group -- section
+
+section add_comm_group
+
+variables {γ : Type*}
+  [topological_space γ] [second_countable_topology γ] [add_comm_group γ] [topological_add_group γ]
+
+instance : add_comm_group (α →ₘ γ) :=
+{ add_comm := ae_eq_fun.add_comm
+  .. ae_eq_fun.add_group
+}
+
+end add_comm_group -- section
+
+section semimodule
+
+variables {K : Type*} [semiring K] [topological_space K] [second_countable_topology K]
+variables {γ : Type*} [topological_space γ] [second_countable_topology γ]
+          [add_comm_monoid γ] [semimodule K γ] [topological_semimodule K γ]
+
+protected def smul : K → (α →ₘ γ) → (α →ₘ γ) :=
+λ c f, comp (has_scalar.smul c) (measurable_smul measurable_id) f
+
+instance : has_scalar K (α →ₘ γ) := ⟨ae_eq_fun.smul⟩
+
+@[simp] lemma smul_mk (c : K) (f : α → γ) (hf) : c • (mk f hf) = mk (c • f) (measurable_smul hf) :=
+rfl
+
+protected lemma one_smul : ∀ (f : α →ₘ γ), (1 : K) • f = f :=
+by { rintros ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, one_smul] }
+
+protected lemma mul_smul : ∀ (x y : K) (f : α →ₘ γ), (x * y) • f = x • y • f :=
+by { rintros x y ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, mul_action.mul_smul x y f], refl }
+
+variables [topological_add_monoid γ]
+
+protected lemma smul_add : ∀ (x : K) (f g : α →ₘ γ), x • (f + g) = x • f + x • g :=
+begin
+  rintros x ⟨f, hf⟩ ⟨g, hg⟩, simp only [quot_mk_eq_mk, smul_mk, mk_add_mk],
+  congr, exact smul_add x f g
+end
+
+protected lemma smul_zero : ∀ (x : K), x • (0 : α →ₘ γ) = 0 :=
+by { intro x, simp only [zero_def, smul_mk], congr, exact smul_zero x }
+
+protected lemma add_smul : ∀ (x y : K) (f : α →ₘ γ), (x + y) • f = x • f + y • f :=
+begin
+  intros x y, rintro ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, mk_add_mk], congr,
+  exact add_smul x y f
+end
+
+protected lemma zero_smul : ∀ f : α →ₘ γ, (0 : K) • f = 0 :=
+by { rintro ⟨f, hf⟩, simp only [quot_mk_eq_mk, smul_mk, zero_def], congr, exact zero_smul K f }
+
+instance : semimodule K (α →ₘ γ) :=
+{ one_smul  := ae_eq_fun.one_smul,
+  mul_smul  := ae_eq_fun.mul_smul,
+  smul_add  := ae_eq_fun.smul_add,
+  smul_zero := ae_eq_fun.smul_zero,
+  add_smul  := ae_eq_fun.add_smul,
+  zero_smul := ae_eq_fun.zero_smul }
+
+end semimodule -- section
+
+section vector_space
+
+variables {K : Type*} [discrete_field K] [topological_space K] [second_countable_topology K]
+variables {γ : Type*} [topological_space γ] [second_countable_topology γ] [add_comm_group γ]
+          [topological_add_group γ] [vector_space K γ] [topological_semimodule K γ]
+
+instance : vector_space K (α →ₘ γ) := { .. ae_eq_fun.semimodule }
+
+end vector_space -- section
+
+open ennreal
+-- integral on ae_eq_fun
+def integral (f : α →ₘ nnreal) : ennreal :=
```

Comment:
This troubles me a lot, too.  It would be good if everything can be written in `ennreal`. However, there is no `enorm`. So how do I define integrable functions? One way is
```lean
def integrable (f : α → γ) : Prop := (∫⁻ a, ennreal.of_real (norm (f a))) < ⊤
```
But I find dealing with `ennreal.of_real` troublesome, and in anyways, `ennreal.of_real r` is defined to be `coe (nnreal.of_real r)`, i.e., first transform `real` to `nnreal`, then transform `nnreal` to `ennreal`. So I think, why not use `nnnorm` instead. Hence, I defined integrable functions as
```lean
def integrable (f : α → γ) : Prop := (∫⁻ a, nnnorm (f a)) < ⊤
```
I guess that's why I let this `integral` to be on functions into `nnreal`.

There is no `enormed_space`. The function being integrated is always `α → normed_space`, so it seems that lifting `metric_space` to `emetric_space` is enough.  I understand that `metric_space` is also `emetric_space`, and so lifting `metric_space` to `emetric_space` can be seen as an instance of lifting `emetric_space` to `emetric_space`. But again, `edist` of a `metric_space` is defined to be `ennreal.of_real (dist x y)`, and `ennreal.of_real` is really just one more step of things.

Reply:
I agree with Johannes that this should definitely be defined for function taking values in `ennreal`. Integrable functions come later. You could require that the eintegral of `nnorm ...` should be finite (and a coercion would fire up automatically to interpret this as an `ennreal`). It would probably be even better to define an `enorm` and use it (and probably `enormed_space`, but this can and will wait).

Reply:
Now that I have looked at the rest of the file, I am sure that you should define `enorm` and use it most often.

Reply:
I get the same feeling that using `enorm` is the most natural thing. I even have a file working on `enormed_group` and `enormed_space`. But it is not complete. I just copied the entire `normed_space.basic` file and changed all `norm` to `enorm`. Are there anything special about `enormed_space`?

Reply:
@johoelzl @sgouezel I discussed this with Joe yesterday. Currently we have (among other things)
```lean
lintegral (f : α → ennreal) : ennreal
integral (f : α →ₘ nnreal) : ennreal
l1.bochner_integral [normed_space ℝ γ] (f : α →₁ γ)
bochner_integral [normed_space ℝ γ] (f : α → γ)
```
At the end of the day, we only care about the last two, right? The others are used to construct the Bochner integral and derive basic properties, but end users should only use Bochner. If so, I think the last two should be named `integral`. The others could have more descriptive names, but otherwise they seem irrelevant.

Anyhow, you want:
```lean
eintegral (f : α →ₘ ennreal) : ennreal
```
As a stepping-stone to the Bochner integral, this doesn't seem to make things any easier, since (as Joe says), later on he will have to deal with casts from an `nnreal` to an `ennreal`. Is `eintegral` useful on its own? Or can you say why you prefer this approach?

Reply:
You have two options:
* For now, just define a function `enorm` on normed spaces, and work with it. With the possibility to come back later to this and extend everything to enormed spaces.
* Or define now enormed spaces and prove that there is an instance from normed spaces to enormed spaces. To an enormed space is also canonically associated a normed space (the set of elements with finite enorm). This will show up a lot in analysis. For instance, one could define `L^1` as an enormed space containing all (equivalence classes of) measurable functions, and the associated normed space would be the space of integrable functions.

For this PR, I think it is wiser to just go for the first option. Otherwise, it will become too big.

Reply:
I wrote my comment before Joe's arrived. So it sounds like the proposal is:
- define `enormed_group` and `enormed_space` and prove basic properties. (It is probably best to do this as a separate PR.)
- rebase this PR on that, and define `eintegral (f : α →ₘ ennreal) : ennreal`
Then in the end we will have `ebochner_integral` and `bochner_integral`, which I think really ought to be renamed to `eintegral` and `integral`.
@johoelzl @sgouezel can you confirm that you think this is the right approach?

Reply:
Sorry, now my message crossed Sébastien's. Both approaches sound good to me.

Reply:
I agree that `integral (f : α →ₘ nnreal) : ennreal` should be *replaced* by `integral (f : α →ₘ ennreal) : ennreal`. I see no reason to have both, and when dealing with an nnreal function the coercion should make this relatively painless. (It's also not a horrible patch map like the ones between ennreal and real, so it should not be a big problem to manipulate.) There is no `ebochner_integral`; bochner requires a real vector space, and there is no extended real analogue.

Reply:
Mario is right that there is no `ebochner_integral`. I think the terminology should be `eintegral (f : α →ₘ ennreal) : ennreal`, and what is currently called `bochner_integral` should just be `integral`.

Reply:
I have a version that uses the `integral (f : α →ₘ ennreal) : ennreal`. Note that this `integral` is used to define the `edist` between two `ae_eq_fun`s. I list the relevant definitions and lemmas below. The important one is the last one.

```lean
def mk (f : α → β) (hf : measurable f) : α →ₘ β := quotient.mk ⟨f, hf⟩

def integral (f : α →ₘ ennreal) : ennreal :=
quotient.lift_on f (λf, lintegral f.1) (assume ⟨f, hf⟩ ⟨g, hg⟩ eq, lintegral_congr_ae eq)

lemma integral_mk (f : α → ennreal) (hf) : integral (mk f hf) = lintegral f :=
rfl

variables {γ : Type*} [emetric_space γ] [second_countable_topology γ]

def comp_edist (f g : α →ₘ γ) : α →ₘ ennreal := comp₂ edist measurable_edist' f g

def edist (f g : α →ₘ γ) : ennreal := integral (comp_edist f g)

lemma edist_mk_mk {f g : α → γ} (hf hg) :
  edist (mk f hf) (mk g hg) = ∫⁻ x, edist (f x) (g x) := rfl
```
Now if `γ` is an `emetric_space` (or an `enormed_space`), then everything fits perfectly. But in the end `γ` is a `normed_space` and so is a `metric_space`, and so 
```lean
edist (mk f hf) (mk g hg) = ∫⁻ x, edist (f x) (g x) 
... = ∫⁻ x, metric_space.edist (f x) (g x)
... = ∫⁻ x, ennreal.of_real (dist (f x) (g x))
```
I don't see how we can avoid `metric_space.edist`, or `ennreal.of_real`. There's no supporting lemma for `metric_space.edist` so everything needs to be proved from the start. 

Now if we use the `nnreal` version:
```lean
def integral (f : α →ₘ nnreal) : ennreal :=
quotient.lift_on f (λf, ∫⁻ a, f.1 a)
  (assume ⟨f, hf⟩ ⟨g, hg⟩ eq, lintegral_congr_ae $ by simpa [coe_eq_coe])

@[simp] lemma integral_mk (f : α → nnreal) (hf) : integral (mk f hf) = ∫⁻ a, f a :=
rfl

variables {γ : Type*} [metric_space γ] [second_countable_topology γ]

def comp_nndist (f g : α →ₘ γ) : α →ₘ nnreal := comp₂ nndist measurable_nndist' f g

protected def edist (f g : α →ₘ γ) : ennreal := integral (comp_nndist f g)

lemma edist_mk_mk {f g : α → γ} (hf hg) :
  edist (mk f hf) (mk g hg) = ∫⁻ x, nndist (f x) (g x) := rfl
```
This is nicer because here I only need to deal with the coercion from `nnreal` to `ennreal`, which is relatively easy to do, unlike the one above.

The disadvantage is that when `γ` is an `emetric_space`, `edist` of two `ae_eq_fun`s is not defined. But right now we don't have `enormed_space`, so the case where `γ` is an `emetric_space` will not appear at all.

Reply:
So I would agree with you if `enormed_space` had already been set up nicely. But without an `enormed_space`, instead of switching from `emetric_space` to `metric_space` and `normed_space` in the middle, I suggest we commit to `metric_space` from the start.

Reply:
Well, `edist` can be expressed in terms of a coercion from an`nnreal`:
```lean
section
variables {α : Type*} [metric_space α]

theorem edist_nndist (x y : α) : edist x y = nndist x y :=
by { rw [edist_dist, nndist, ennreal.of_real, nnreal.of_real], congr, 
     rw [max_eq_left dist_nonneg] }

end
```
Then it sees that in the third line of your calculation you can rewrite the new `edist` to exactly the one you want. In other words, you can still prove the `edist_mk_mk` you want, even though the proof is not `rfl`. Does that work?

Reply:
In fact, now I see that `edist_nndist` is already in the library. (I wrote it as an `example` when I was trying it out, so Lean didn't complain.) It might make sense to add the corresponding identities to `ennreal.lean`:
```lean
namespace ennreal

lemma coe_nnreal_eq (r : nnreal) : 
  (r : ennreal) = ennreal.of_real r :=
by { rw [ennreal.of_real, nnreal.of_real], cases r with r h, congr, dsimp, rw max_eq_left h }

lemma of_real_eq_coe_nnreal {x : real} (h : x ≥ 0) :
  ennreal.of_real x = @coe nnreal ennreal _ (⟨x, h⟩ : nnreal) :=
by { rw [coe_nnreal_eq], refl }

end ennreal

section
variables {α : Type*} [metric_space α]

example (x y : α) : edist x y = nndist x y :=
by { rw [edist_dist, nndist, ennreal.of_real_eq_coe_nnreal] }

end
```